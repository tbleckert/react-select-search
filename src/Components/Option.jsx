import { memo } from 'react';
import PropTypes from 'prop-types';
import classes from '../lib/classes';

const Option = ({
    optionProps,
    highlighted,
    selected,
    option,
    cls,
    renderOption,
}) => {
    const props = { ...optionProps, value: option.value, disabled: option.disabled };
    const className = classes({
        [cls('option')]: true,
        [cls('is-selected')]: selected,
        [cls('is-highlighted')]: highlighted,
    });

    return (
        <li className={cls('row')} role="menuitem" data-index={option.index} data-value={escape(option.value)} key={option.value}>
            {renderOption && renderOption(props, option, { selected, highlighted }, className)}
            {!renderOption && <button type="button" className={className} {...props}>{option.name}</button>}
        </li>
    );
};

Option.defaultProps = {
    renderOption: null,
};

Option.propTypes = {
    option: PropTypes.shape({
        name: PropTypes.string.isRequired,
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        disabled: PropTypes.bool,
        index: PropTypes.number,
    }).isRequired,
    highlighted: PropTypes.bool.isRequired,
    selected: PropTypes.bool.isRequired,
    optionProps: PropTypes.shape({
        tabIndex: PropTypes.string.isRequired,
        onMouseDown: PropTypes.func.isRequired,
    }).isRequired,
    cls: PropTypes.func.isRequired,
    renderOption: PropTypes.func,
};

export default memo(Option);
