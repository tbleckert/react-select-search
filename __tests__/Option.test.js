import renderer from 'react-test-renderer';
import Option from '../src/Components/Option';

const optionProps = {
    tabIndex: '-1',
    onMouseDown: () => {},
};

describe('Test Option component', () => {
    test('Renders with default props', () => {
        const component = renderer.create(
            <Option
                option={{
                    name: 'Hamburger',
                    value: 'hamburger',
                }}
                cls={(cls) => cls}
                highlighted={false}
                selected={false}
                optionProps={optionProps}
            />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Highlighted', () => {
        const component = renderer.create(
            <Option
                option={{
                    name: 'Hamburger',
                    value: 'hamburger',
                }}
                cls={(cls) => cls}
                highlighted={true}
                selected={false}
                optionProps={optionProps}
            />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Selected', () => {
        const component = renderer.create(
            <Option
                option={{
                    name: 'Hamburger',
                    value: 'hamburger',
                }}
                cls={(cls) => cls}
                highlighted={false}
                selected={true}
                optionProps={optionProps}
            />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
