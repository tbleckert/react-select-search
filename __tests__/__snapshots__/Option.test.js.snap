// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Option component Highlighted 1`] = `
<li
  className="row"
  data-value="hamburger"
  role="menuitem"
>
  <button
    className="option is-highlighted"
    onMouseDown={[Function]}
    tabIndex="-1"
    type="button"
    value="hamburger"
  >
    Hamburger
  </button>
</li>
`;

exports[`Test Option component Renders with default props 1`] = `
<li
  className="row"
  data-value="hamburger"
  role="menuitem"
>
  <button
    className="option"
    onMouseDown={[Function]}
    tabIndex="-1"
    type="button"
    value="hamburger"
  >
    Hamburger
  </button>
</li>
`;

exports[`Test Option component Selected 1`] = `
<li
  className="row"
  data-value="hamburger"
  role="menuitem"
>
  <button
    className="option is-selected"
    onMouseDown={[Function]}
    tabIndex="-1"
    type="button"
    value="hamburger"
  >
    Hamburger
  </button>
</li>
`;
